vignettes/namespaces.Rmd

```diff
@@ -0,0 +1,416 @@
+
+```{r}
+cat(xslt::xml_xslt(xml, sty))
+```
+
+We can in this way programatically transform the content of the markdown. In 
+this example, we can change the `**test**` to be an inline R code chunk that
+emits `_test_`. 
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml()
+
+xml2::xml_find_all(xml, "//md:strong", ns) %>%
+  xml2::xml_set_name("code") %>%
+  xml2::xml_set_text("r cat('_test_')")
+
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+### Perils: adding nodes
+
+A default namespace is all fun and games until you need to add new nodes. Take
+for example the situation where we want to add a code block. In commonmark, it's
+a `code_block` node with an `info` attribute stating the language and the text
+inside is the code.
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() 
+xml2::xml_add_child(xml, "code_block", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//md:code_block", ns)
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+By all means, the node should have added correctly, but because we did not 
+specify a namespace, *it is not recognized as part of the `md` namespace* even
+though we added it as a child of the document. The [best way to handle this
+situation is to reparse the document][reparse]:
+
+```{r reread}
+xml %>%
+  as.character() %>%
+  xml2::read_xml() %>%
+  xslt::xml_xslt(sty) %>%
+  cat()
+```
+
+We *could* also try adding the namespace to the node when we add it:
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() 
+xml2::xml_add_child(xml, "code_block", 
+  xmlns = "http://commonmark.org/xml/1.0", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//md:code_block", ns)
+cat(xslt::xml_xslt(xml, sty))
+```
+
+It works, but let's take a look at our namespaces:
+
+```{r}
+xml2::xml_ns(xml)
+```
+
+Every node we add with an unnamed namespace adds another default and in the end, 
+if we are doing a lot of substitution, we can end up with hundreds of namespaces.
+
+
+## No Namespace?
+
+What if we just tried to use no namespace?
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() %>%
+  xml2::xml_ns_strip()
+xml2::xml_add_child(xml, "code_block", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//code_block")
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+We can now add new nodes and use XPath without namespace prefixes or objects,
+but we have lost the ability to use our stylesheet :(
+
+But! Maybe we can do this by adding the namespace at the last minute!
+
+```{r}
+xml2::xml_set_attr(xml, "xmlns", "http://commonmark.org/xml/1.0")
+xml
+cat(xslt::xml_xslt(xml, sty))
+```
+
+## Harnessing the power of namespaces
+
+When you know that namespaces with prefixes will only respond to nodes with that
```

Comment:
Add a practical use case (in words, not code necessarily) for this?

Reply:
Good question! I use the alternate namespace in {pegboard} to help me identify and label pandoc fenced-div sections by adding pairs of equivalently labeled tags that are not part of the markdown document so I can easily parse the content with `find_between()`

Otherwise, I can see the masking pattern useful if you wanted to create several versions of the same prose in a single document (e.g. if you were creating a quiz that you wanted randomized per student).